### PR TITLE
Hardware Key docs

### PR DESCRIPTION
**Pull request by Joerger**
Adds new docs for [Hardware Key Support](https://github.com/gravitational/teleport/pull/14774)

### CODE REVIEW

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+```yaml
+kind: role
+metadata:
+  name: admin
+spec:
+  allow:
+    require_session_mfa: hardware_key_touch
+```
+
+You can also enabled Hardware Key support cluster-wide by updating yout Teleport configuration as below:
```
Comment (strideynet) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:77]:
```suggestion
You can also enabled Hardware Key support cluster-wide by updating your Teleport configuration as below:
```

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+[Per-session MFA](per-session-mfa.mdx) provides a way to require an MFA presence check with a user's login session, but this check is only used for starting sessions with Teleport services (SSH, Kube, DB, etc.). This means that non-session actions, like administrative commands run with `tctl`, can still be run with an exfiltrated login session. 
+
+To prevent these types of exfiltration attacks altogether, Teleport clients support hardware-based private keys. Unlike disk-based private keys, hardware-based private keys are generated and stored directly on a hardware device and are impossible to export. With this features, a user's login session is only functional with access to the user's hardware device.
+
+Additionally, this feature can be configured to require touch for every Teleport request, including non-session requests. With touch required, Hardware Key Support provides better security in comparison to our per-session MFA offering.
+
+<Admonition type="note" title="touch caching">
+
+  A user's touch is cached on the YubiKey for 15 seconds to prevent excessive touch prompts.
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:18]: This is the first time you mention YubiKeys specifically. Is the caching only supported for YubiKeys, or does this apply to all supported hardware keys?

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+</Admonition>
+
+<ScopedBlock scope="oss">
+
+  Hardware-backed Login Sessions require Teleport Enterprise or Teleport Cloud.
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:24]: Teleport Cloud _is_ Teleport Enterprise. Let's confirm with @xinding33 how we should phrase this.

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+
+<Admonition type="note" title="PIV support">
+
+  Hardware-backed Login Sessions require a PIV-compatible hardware key. Currently, this feature is only guarenteed to support YubiKey series 5+.
+
+</Admonition>
+
+<Admonition type="warning" title="Compatibility Issues">
+
+  Hardware-backed Login Sessions provides the best security available, but it is a new feature with some compatiblity trade-offs. We hope to remove these incompatibilities as development progresses.
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:36]:
```suggestion
  Hardware-backed Login Sessions provides the best security available, but it is a new feature with some compatiblity trade-offs. We hope to improve compatibility in the future.
```

Let's phrase this in a way to discuss what we're adding in the future, not what we're removing.

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+  
+  Currently, Hardware-backed Login Sessions support `tsh`, `tctl`, and Teleport Connect for:
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:38]: This works with Teleport Connect without additional code? If so, nice!

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+
+  To navigate these incompatiblities, we recommend enabled Hardware Key support only when necessary, such as for roles with access to critical infrastructure. These roles can be accessed as needed with Access Requests so that users can avoid these issues for their normal login sessions.
+
+</Admonition>
+
+## Prerequisites
+
+(!docs/pages/includes/edition-prereqs-tabs.mdx!)
+
+- A series 5+ YubiKey
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:60]: If we don't depend on anything specific to YubiKeys, I would rephrase this to make it less likely to go out of date if other vendors release compatible keys.
```suggestion
- A hardware key that supports PIV
```

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+## Step 2/2. Log in
+
+Once enabled on the Auth server, affected users will be prompted to connect and touch their YubiKey on login:
+
+<ScopedBlock scope={["oss", "enterprise"]}>
+
+```code
+$ tsh login --user=dev --proxy=proxy.example.com:3080
+# Enter password for Teleport user dev:
+# Tap your YubiKey
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:165]: Similar to above comments - did we actually hard-code "YubiKey" in the source code? Should we have?

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+#   Roles:              access, editor
+#   Logins:             bjoerger
+#   Kubernetes:         enabled
+#   Valid until:        2022-10-11 01:53:44 -0700 PDT [valid for 8h0m0s]
+#   Extensions:         permit-X11-forwarding, permit-agent-forwarding, permit-port-forwarding, permit-pty, private-key-policy
+```
+
+</ScopedBlock>
+
+Affected user's with existing sessions will be prompted to re-login on their next request:
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:196]:
```suggestion
Affected users with existing sessions will be prompted to re-login on their next request:
```

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+
+Hardware Key support is disbled by default. You can enable Hardware Key support for specific roles like below:
+
+```yaml
+kind: role
+metadata:
+  name: admin
+spec:
+  allow:
+    require_session_mfa: hardware_key_touch
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:74]: Please also update the role reference page to list the valid values for `require_session_mfa`. I'm pretty sure it used to be a boolean, if that's changed we should highlight it and document all possible values.

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+# Tap your YubiKey
+# Cluster Name Status Cluster Type Labels Selected 
+# ----------- ------ ------------ ------ -------- 
+# example.com online root                *        
+```
+
+## Configuring Hardware Key support with per-session MFA
+
+For existing users of per-session MFA, upgrading to `hardware_key_touch` may be disruptive, as it requires touch on all requests as opposed to only session requests. While this upgrade would yield the best security, we can get a very similar level of security by combining Hardware Key support with per-session MFA.
+
```
Comment (zmb3) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:212]: So we support both `hardware_key` and `hardware_key_touch` and they behave differently?

You have to "touch" a hardware key in both cases, right? I find the choice of names here pretty confusing.

Comment (Joerger) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:212]: I agree, the confusion comes from the per-session MFA touch being completely separate from PIV touch. I think there's an argument for removing `hardware_key` altogether, and only having `hardware_key_touch` (probably changing the name to `hardware_key`), but it seems a bit late to do that.

Ultimately this is the solution we were able to agree on after some [back and forth in the RFD](https://github.com/gravitational/teleport/pull/14774#issuecomment-1218306104).

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 2 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:18]: Currently, we only really support YubiKeys. there aren't any other (significant) PIV-compatible smart cards out there. See this section of the RFD covering this - https://github.com/gravitational/teleport/blob/cf595d8b0feeea863669d9e258186396ef225cb6/rfd/0080-piv-private-key.md#library

The touch cache is one of the things that may not be supported in other smart card implementations in the future.

I'll move this block below the PIV support block below, which is meant to clear up the change of wording from "hardware key" to "YubiKey".

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 5 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:38]: Well there was some additional code to get it working, and some additional code needed to add the [touch prompt](https://github.com/gravitational/teleport/issues/17311). Currently Teleport Connect users would just see their key blinking.

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 6 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:60]: We will need to update the implementation and docs if additional vendors add PIV compatible keys, see the comment above.

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 9 above)
Comment (Joerger) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:74]: I don't see a place to list the valid values, but I updated the references's description and value type.

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 6 above)
Comment (zmb3) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:60]: Got it, thanks.

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 10 above)
Comment (jakule) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:212]: Yes, we don't want users to enable MFA and PIV at the same time.

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 7 above)
Comment (jakule) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:165]: As Brian mentioned above. We want to use YubiKey in user messages, as PIV is not a very well-known standard. Currently, YubiKey is the only popular key that uses it. We also didn't test anything else.

File: docs/pages/access-controls/guides/hardware-key-support.mdx (hunk 10 above)
Comment (zmb3) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:209]: Sure but is it too late to at least have better names? Both of them sound the same to me.

Comment (Joerger) [reply docs/pages/access-controls/guides/hardware-key-support.mdx:209]: For some background, the names originally were made for a new field: `private_key_policy: none | hardware_key | hardware_key_touch`, which as I mentioned is separate from per-session MFA, but was rejected in RFD review. In the context of the field `require_session_mfa`, neither of these make sense, and I don't know how to make them make sense. I'm really at a loss for any better names, but I'm open to suggestions.

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+# cluster auth preference has been updated
+```
+
+</ScopedBlock>
+
+## Step 2/2. Log in
+
+Once enabled on the Auth server, affected users will be prompted to connect and touch their YubiKey on login:
+
+<ScopedBlock scope={["oss", "enterprise"]}>
```
Comment (jakule) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:157]: So is this feature Ent only or OSS too?

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+
+[Per-session MFA](per-session-mfa.mdx) provides a way to require an MFA presence check with a user's login session, but this check is only used for starting sessions with Teleport services (SSH, Kube, DB, etc.). This means that non-session actions, like administrative commands run with `tctl`, can still be run with an exfiltrated login session. 
+
+To prevent these types of exfiltration attacks altogether, Teleport clients support hardware-based private keys. Unlike disk-based private keys, hardware-based private keys are generated and stored directly on a hardware device and are impossible to export. With this features, a user's login session is only functional with access to the user's hardware device.
+
+Additionally, this feature can be configured to require touch for every Teleport request, including non-session requests. With touch required, Hardware Key Support provides better security in comparison to our per-session MFA offering.
+
+<ScopedBlock scope="oss">
+  Hardware Key Support requires Teleport Enterprise.
```
Comment (jakule) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:21]: Technically, hardware keys should also work with Teleport Cloud?

**Review by jakule — APPROVED** [review]
LGTM

**Comment by alexfornuto** [comment]
I assume this should be backported to branch/11, any other branches?

File: docs/pages/access-controls/guides/hardware-key-support.mdx [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+[Per-session MFA](per-session-mfa.mdx) provides a way to require an MFA presence check with a user's login session, but this check is only used for starting sessions with Teleport services (SSH, Kube, DB, etc.). This means that non-session actions, like administrative commands run with `tctl`, can still be run with an exfiltrated login session. 
+
+To prevent these types of exfiltration attacks altogether, Teleport clients support hardware-based private keys. Unlike disk-based private keys, hardware-based private keys are generated and stored directly on a hardware device and are impossible to export. With this features, a user's login session is only functional with access to the user's hardware device.
+
+Additionally, this feature can be configured to require touch for every Teleport request, including non-session requests. With touch required, hardware key support provides better security in comparison to our per-session MFA offering.
+
+<Admonition type="note" title="PIV support">
+  Hardware key support requires users to use a PIV-compatible hardware key. Currently, this feature is only guarenteed to support YubiKey series 5+.
```
Comment (alexfornuto) [inline docs/pages/access-controls/guides/hardware-key-support.mdx:21]: Perhaps add a line here about what this process will do to any existing keys/certs in the PIV store of the key.

**Review by alexfornuto — APPROVED** [review]
I'm generally 👍, assuming the comments below are addressed.
